File: test/benchmark_end_to_end.jl
using AlphaHull

function halton(i, base)
    value = 0.0
    factor = 1.0 / base
    while i > 0
        value += (i % base) * factor
        i = div(i, base)
        factor /= base
    end
    return value
end

function benchmark_points(n)
    points = Matrix{Float64}(undef, n, 2)
    for i in 1:n
        points[i, 1] = halton(i, 2)
        points[i, 2] = halton(i, 3)
    end
    return points
end

function median_time(f; repeats=7)
    f()
    GC.gc()
    times = sort([@elapsed f() for _ in 1:repeats])
    return times[cld(length(times), 2)]
end

for n in (500, 2000)
    points = benchmark_points(n)
    alpha = 0.07
    elapsed = median_time(() -> begin
        hull = ahull(points; alpha)
        areaahull(hull)
    end)
    println("Julia ahull + areaahull n=$n: $(round(elapsed; digits=4)) s")
end

File: test/benchmark_end_to_end.R
benchmark_points <- function(n) {
  halton <- function(i, base) {
    value <- 0
    factor <- 1 / base
    while (i > 0) {
      value <- value + (i %% base) * factor
      i <- i %/% base
      factor <- factor / base
    }
    value
  }
  i <- seq_len(n)
  cbind(vapply(i, halton, numeric(1), base=2),
        vapply(i, halton, numeric(1), base=3))
}

median_time <- function(f, repeats=7) {
  f()
  times <- replicate(repeats, unname(system.time(f())["elapsed"]))
  median(times)
}

library(alphahull)
for (n in c(500, 2000)) {
  points <- benchmark_points(n)
  alpha <- 0.07
  elapsed <- median_time(function() {
    hull <- ahull(points, alpha=alpha)
    areaahull(hull)
  })
  cat(sprintf("R ahull + areaahull n=%d: %.4f s\n", n, elapsed))
}
